impl/pushchannelmonitor.go

```diff
@@ -0,0 +1,254 @@
+package impl
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	datatransfer "github.com/filecoin-project/go-data-transfer"
+	"github.com/filecoin-project/go-data-transfer/channels"
+)
+
+type monitorAPI interface {
+	SubscribeToEvents(subscriber datatransfer.Subscriber) datatransfer.Unsubscribe
+	RestartDataTransferChannel(ctx context.Context, chid datatransfer.ChannelID) error
+	CloseDataTransferChannel(ctx context.Context, chid datatransfer.ChannelID) error
+}
+
+// pushChannelMonitor watches the data-rate for push channels, and restarts
+// a channel if the data-rate falls too low
+type pushChannelMonitor struct {
+	ctx  context.Context
+	stop context.CancelFunc
+	mgr  monitorAPI
+	cfg  *pushMonitorConfig
+
+	lk       sync.RWMutex
+	channels map[*monitoredChannel]struct{}
+}
+
+type pushMonitorConfig struct {
+	Interval       time.Duration
+	MinBytesSent   uint64
+	RestartBackoff time.Duration
+}
+
+func newPushChannelMonitor(mgr monitorAPI) *pushChannelMonitor {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &pushChannelMonitor{
+		ctx:      ctx,
+		stop:     cancel,
+		mgr:      mgr,
+		channels: make(map[*monitoredChannel]struct{}),
+	}
+}
+
+func (m *pushChannelMonitor) setConfig(cfg *pushMonitorConfig) {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	m.cfg = cfg
+}
+
+// add a channel to the push channel monitor
+func (m *pushChannelMonitor) add(chid datatransfer.ChannelID) *monitoredChannel {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	mpc := newMonitoredChannel(m.mgr, chid, m.cfg, m.onMonitoredChannelShutdown)
+	m.channels[mpc] = struct{}{}
+	return mpc
+}
+
+func (m *pushChannelMonitor) shutdown() {
+	m.stop()
+}
+
+// onShutdown shuts down all monitored channels
+func (m *pushChannelMonitor) onShutdown() {
+	m.lk.RLock()
+	defer m.lk.RUnlock()
+
+	for ch := range m.channels {
+		ch.shutdown()
+	}
+}
+
+// onMonitoredChannelShutdown is called when a monitored channel shuts down
+func (m *pushChannelMonitor) onMonitoredChannelShutdown(mpc *monitoredChannel) {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	delete(m.channels, mpc)
+}
+
+func (m *pushChannelMonitor) enabled() bool {
+	return m.cfg != nil && m.cfg.Interval > 0
+}
+
+func (m *pushChannelMonitor) start() {
+	go m.run()
+}
+
+func (m *pushChannelMonitor) run() {
+	defer m.onShutdown()
+
+	// Check the data-rate of all monitored channels on every tick
+	var tickerChan <-chan time.Time
+	if m.enabled() {
```

Comment:
Shouldn't we exit `run()` if `m.enabled()` is `false`?

Reply:
agree on the "enabled" design being a little wonky -- presumably if any point someone calls setConfig(), then "enabled" becomes true, right?

but based on the code as written, it's not actually enabled, cause the Ticker never tics.

I feel like:
- move towards clarity that enabled actually can't be changed (would help if this were it's own package, and enabled was a private function where Start & SetConfig were not)
- allow calling setconfig to actually enable the monitor even is start was called earlier